core/src/main/scala/com/pingcap/tispark/TiBatchWrite.scala

```diff
@@ -304,15 +299,52 @@ class TiBatchWrite(@transient val df: DataFrame,
       }
     }
 
-    val deduplicatedTiRowRdd = deduplicateIfNecessary(tiRowRdd, tiTableInfo.isPkHandle)
-
     // get timestamp as start_ts
     val startTimeStamp = tiSession.getTimestamp
 
-    // TODO support insert ignore
-    val toBeCheckedRdd = getKeysNeedCheck(deduplicatedTiRowRdd, startTimeStamp)
-    checkConflictWithInsertKeys(toBeCheckedRdd)
-    val encodedTiRowRDD = generateRDDToBeInserted(toBeCheckedRdd)
+    // for partition table, we need calculate each row and tell which physical table
+    // that row is belong to.
+    // currently we only support upsert and insert.
+    val constraintCheckIsNeeded = handleCol != null || uniqueIndices.nonEmpty
+
+    val encodedTiRowRDD = if (constraintCheckIsNeeded) {
+      val wrappedRowRdd = if (tiTableInfo.isPkHandle) {
+        tiRowRdd.map { row =>
+          WrappedRow(row, extractHandleId(row))
+        }
+      } else {
+        val start = getAutoTableIdStart(tiRowRdd.count)
+        tiRowRdd.zipWithIndex.map { data =>
+          WrappedRow(data._1, data._2 + start)
+        }
+      }
+
+      val distinctWrappedRowRdd = deduplicate(wrappedRowRdd)
+
+      val rddDeletion = generateDataToBeRemovedRdd(distinctWrappedRowRdd, startTimeStamp)
+      if (!options.replace && !rddDeletion.isEmpty()) {
+        throw new TiBatchWriteException("data to be inserted has conflicts with TiKV data")
+      }
+      val mergedRDD = generateKV(distinctWrappedRowRdd, remove = false) ++ generateKV(rddDeletion, remove = true)
+
+      mergedRDD.groupByKey().map {
+        case(key, iterable) =>
+          // if rdd contains same key, it means we need delete and update the value associated the
```

Comment:
would u pls give an example which will cause `iterable.size >= 2` ?

Reply:
We discussed f2f.